### ISSUE TITLE
Tests take hours under ASAN-instrumented perl

### ISSUE THREAD

**Issue by ribasushi** [ingydotnet/yaml-pm#186]
[This block](https://metacpan.org/source/TINITA/YAML-1.24/t/2-scalars.t#L31-35) and [also this one](https://metacpan.org/source/TINITA/YAML-1.24/t/rt-90593.t#L12-18) take literally hours to complete with 100% CPU activity on perls instrumented as follows:
`perlbrew install -v -j10 --as 5.24.3_asan -Dusethreads -Accflags="-fsanitize=address -g -O2" -Aldflags="-fsanitize=address -g -O2" perl-5.24.3`

Please consider disabling these tests based on examining `$Config{ccflags}`

Alternatively optimize YAML.pm to not require a ton of memory allocations in these cases.

**Comment by perlpunk** [comment]
Thanks. Could you maybe try out if this patch mentioned in issue #187 resolves the problem?
https://github.com/ingydotnet/yaml-pm/issues/187#issuecomment-374769416

**Comment by perlpunk** [comment]
I wanted to try it myself, but install fails if I call perlbrew with these options:
```
SUMMARY: AddressSanitizer: 240600 byte(s) leaked in 1154 allocation(s).
Failed to build miniperl. Please run make minitest
makefile:382: recipe for target 'lib/buildcustomize.pl' failed
make: *** [lib/buildcustomize.pl] Error 1
make: *** Waiting for unfinished jobs....
op.c: In function ‘S_fold_constants’:
op.c:4272:28: warning: argument ‘o’ might be clobbered by ‘longjmp’ or ‘vfork’ [-Wclobbered]
 S_fold_constants(pTHX_ OP *o)
                            ^
5.24.3_asan is successfully installed.
```
Although the last line says that it was installed, I can't find it anywhere...

**Comment by perlpunk** [comment]
I created a PR: #188 
Would be cool if you could test it.
`prove -lr test`

**Comment by ribasushi** [comment]
> I created a PR: #188
> Would be cool if you could test it.

It is orders of magnitude better, but still a bit too long. I still recommend skipping the test under debugging configs:
```
rabbit@Ahasver:/dev/shm/yaml-pm$ prove -l test/2-scalars.t 
test/2-scalars.t .. ok          
All tests successful.
Files=1, Tests=10, 11407 wallclock secs ( 0.10 usr  0.05 sys + 6110.04 cusr 5186.05 csys = 11296.24 CPU)
Result: PASS
```
```
rabbit@Ahasver:/dev/shm/yaml-pm$ git checkout perlpunk/regex-quoted
Switched to branch 'perlpunk/regex-quoted'
Your branch is up-to-date with 'origin/perlpunk/regex-quoted'.
```
```
rabbit@Ahasver:/dev/shm/yaml-pm$ prove -l test/2-scalars.t 
test/2-scalars.t .. ok    
All tests successful.
Files=1, Tests=12, 103 wallclock secs ( 0.10 usr  0.01 sys + 76.34 cusr 25.56 csys = 102.01 CPU)
Result: PASS
```

> I wanted to try it myself, but install fails if I call perlbrew with these options:

Don't know what to tell you: it works out of the box for me precisely as I wrote in the original report. Given these perl versions are considered "supported" ( fsvo ), perhaps ask on #p5p why it doesn't work for you. My binutils/gcc are:
```
rabbit@Ahasver:/dev/shm/yaml-pm$ ld -v
GNU ld (GNU Binutils for Debian) 2.28
```
```
rabbit@Ahasver:/dev/shm/yaml-pm$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.9/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 4.9.2-10' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --with-arch-32=i586 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.9.2 (Debian 4.9.2-10) 
```

**Comment by perlpunk** [comment]
Thanks for testing.
What exactly should I check for in ccflags? Sorry for the stupid question, I'm not familiar with the flags.

**Comment by ribasushi** [comment]
> Sorry for the stupid question...

No worries, I wasn't familiar with any of this up until a week ago either.

You want something like this:
`... if $Config{config_args} =~ / \-fsanitize \= (?: address | undefined ) \b /x`

Also you might want to [integrate the entire thing into your CI](https://github.com/Perl5/DBIx-Class/commit/29c01d8351752fac2661d705f484624a16233ab9#diff-354f30a63fb0907d4ad57269548329e3) ( especially for the YAML XS library ). It's been [pretty useful](https://travis-ci.org/Perl5/DBIx-Class/jobs/357739323#L1552-L1660) to me.

**Comment by perlpunk** [comment]
Thanks. Greetings from the Perl Toolchain Summit. I got it compiling with the help of @Tux.
However, now I'm getting many leak errors even when executing the simplest perl script. So I could add this check to the test but never see myself if it's working.
Maybe I need to tweak the compilation flags a bit more.

**Comment by ribasushi** [comment]
> now I'm getting many leak errors even when executing the simplest perl script

Known issue: https://code.activestate.com/lists/perl5-porters/240338/

You need to exactly replicate my original test setup. I am sorry I didn't directly link you to the necessary spots, I was too distracted at the time. Specifically you are missing this: https://github.com/Perl5/DBIx-Class/commit/2e629b86f1e83b25e9ab92747ffaac7b87155693#diff-354f30a63fb0907d4ad57269548329e3R225

Cheers

**Comment by perlpunk** [comment]
Created #198 
I'm too stupid to compile a perl that can reproduce this, so I just added a check to the test.

**Comment by perlpunk** [comment]
Released YAML 1.24_002
